torchao/quantization/README.md

```diff
@@ -164,6 +164,125 @@ model = torch.compile(model, mode='max-autotune')
+
+### Quantization Flow
+What we need to do afterwards is roughly the following
+
+```
+for n, m in model.named_modules():
+    # or use some filter_fn
+    if isinstance(m, torch.nn.Linear):
+        # optional filtering for module name, shape etc.
+        # quantization activation (needed by dynamic quantization)
+        # m.weight = nn.Parameter(to_laq(m.weight, device=..., layout=..., ...))
+        m.weight = nn.Parameter(to_aq(m.weight, device=..., layout=..., ...))
+```
+The model/tensor subclass should also be compatible with AOTI and torch.export, currently we can support
+`torch.export.export` and `torch.aot_compile` with the following workaround:
+```
+from torchao.quantization.utils import unwrap_tensor_subclass
+m_unwrapped = unwrap_tensor_subclass(m)
+
+
+# export
+m = torch.export.export(m_unwrapped, example_inputs).module()
+
+# aot_compile
+torch._export.aot_compile(m_unwrapped, example_inputs)
+```
+
+But we expect this will be integrated into the export path by default in the future.
+
+
+### Example
+Let's use int4 weight only quantization that's targeting tinygemm int4 weight only quantized matmul
+as an example:
+```python
+import torch
+from torchao.quantization.quant_api import quantize
+from torchao.quantization.quant_primitives import MappingType, ZeroPointDomain
+from torchao.dtypes import to_aq
+from torch._inductor.runtime.runtime_utils import do_bench_gpu
+import copy
+
+class ToyLinearModel(torch.nn.Module):
+    def __init__(self, m=64, n=32, k=64):
+        super().__init__()
+        self.linear1 = torch.nn.Linear(m, n, bias=False)
+        self.linear2 = torch.nn.Linear(n, k, bias=False)
+
+    def example_inputs(self, batch_size=1, dtype=torch.float32, device="cpu"):
+        return (torch.randn(batch_size, self.linear1.in_features, dtype=dtype, device=device),)
+
+    def forward(self, x):
+        x = self.linear1(x)
+        x = self.linear2(x)
+        return x
+
+# weight settings
+groupsize = 32
+mapping_type = MappingType.ASYMMETRIC
+block_size = (1, groupsize)
+target_dtype = torch.int32
+quant_min = 0
+quant_max = 15
+eps = 1e-6
+preserve_zero = False
+zero_point_dtype = torch.bfloat16
+zero_point_domain = ZeroPointDomain.FLOAT
+
+dtype = torch.bfloat16
+m = ToyLinearModel(1024, 1024, 1024).eval().to(dtype).to("cuda")
+m_bf16 = copy.deepcopy(m)
+example_inputs = m.example_inputs(dtype=dtype, device="cuda")
+
+m_bf16 = torch.compile(m_bf16, mode='max-autotune')
+
+def apply_weight_quant(weight):
+    return to_aq(weight, mapping_type, block_size, target_dtype, quant_min, quant_max, eps, zero_point_dtype=zero_point_dtype, preserve_zero=preserve_zero, zero_point_domain=zero_point_domain)
+
+m = quantize(m, apply_weight_quant)
+
+torch._inductor.config.force_fuse_int_mm_with_mul = True
+torch._inductor.config.use_mixed_mm = True
+
+# compile the model to improve performance
+m = torch.compile(m, mode='max-autotune')
+
+
+# benchmark to see the speedup
+from torch.utils.benchmark import Timer
```

Comment:
We have a benchmark util now https://github.com/pytorch/ao/blob/main/torchao/utils.py#L4

Mind if we update it in case yours is subtly different?

Reply:
sure, but we probably need to follow up and enable benchmark for cpu and mps as well for that, created an issue here: https://github.com/pytorch/ao/issues/287